datafusion/common/src/dfschema.rs

```diff
@@ -106,37 +106,175 @@ pub type DFSchemaRef = Arc<DFSchema>;
+        self.field_qualifiers
+            .iter()
+            .zip(self.schema.fields().iter())
+            .map(|(qualifier, field)| (qualifier.as_ref(), field))
+    }
+
+    pub fn fields_with_unqualified_name(&self, name: &str) -> Vec<&Field> {
+        self.fields()
+            .iter()
+            .filter(|field| field.name() == name)
+            .map(|f| f.as_ref())
+            .collect()
+    }
+
+    /// Get a list of fields
+    pub fn fields(&self) -> &Fields {
+        &self.schema.fields
+    }
+
+    /// Returns an immutable reference of a specific `Field` instance selected using an
+    /// offset within the internal `fields` vector
+    pub fn field(&self, i: usize) -> &Field {
+        &self.schema.fields[i]
+    }
+
+    /// Returns an immutable reference of a specific `Field` instance selected using an
+    /// offset within the internal `fields` vector and its qualifier
+    pub fn qualified_field(&self, i: usize) -> (Option<&TableReference>, &Field) {
+        (self.field_qualifiers[i].as_ref(), self.field(i))
+    }
+
+    pub fn field_with_qualified_name(
+        &self,
+        qualifier: &TableReference,
+        name: &str,
+    ) -> Option<&Field> {
+        let mut matches = self
+            .iter()
+            .filter(|(q, f)| match q {
+                Some(field_q) => qualifier.resolved_eq(field_q) && f.name() == name,
+                None => false,
+            })
+            .map(|(_, f)| f.as_ref());
+        matches.next()
+    }
+
+    pub fn index_of_column_by_name(
+        &self,
+        qualifier: Option<&TableReference>,
+        name: &str,
+    ) -> Option<usize> {
+        let mut matches = self
+            .iter()
+            .enumerate()
+            .filter(|(_, (q, f))| match (qualifier, q) {
+                // field to lookup is qualified.
+                // current field is qualified and not shared between relations, compare both
+                // qualifier and name.
+                (Some(q), Some(field_q)) => q.resolved_eq(field_q) && f.name() == name,
+                // field to lookup is qualified but current field is unqualified.
+                (Some(_), None) => false,
+                // field to lookup is unqualified, no need to compare qualifier
+                (None, Some(_)) | (None, None) => f.name() == name,
+            })
+            .map(|(idx, _)| idx);
+        matches.next()
+    }
+
+    pub fn field_qualifier(&self, i: usize) -> Option<&TableReference> {
+        self.field_qualifiers[i].as_ref()
+    }
 }
 
 impl DFSchema {
     /// Creates an empty `DFSchema`
     pub fn empty() -> Self {
         Self {
-            inner: Arc::new(Schema::new([])),
-            field_qualifiers: vec![],
+            inner: QualifiedSchema::empty(),
             functional_dependencies: FunctionalDependencies::empty(),
+            metadata: None,
         }
     }
 
+    pub fn metadata_schema(&self) -> &Option<QualifiedSchema> {
```

Comment:
Please add documentation -- imagine you are someone using this API and are not familar with metadata_schema or the content of this API.  I think you would want a short summary of what this is and then a link to the full details